plugins/hashsum_download/server/__init__.py

```diff
@@ -118,7 +123,37 @@ def _getFirstFileByHash(self, algo, hash, user=None):
         return None
 
 
+def _computeHash(event):
+    """
+    Computes all supported checksums on a given file. Downloads the
+    file data and stream-computes all required hashes on it, saving
+    the results in the file document.
+    """
+    file = event.info['file']
+
+    toCompute = SUPPORTED_ALGORITHMS - set(file)
+    toCompute = {alg: getattr(hashlib, alg)() for alg in toCompute}
+
+    if not toCompute:
+        return
+
+    fileModel = ModelImporter.model('file')
+    with fileModel.open(file) as fh:
+        while True:
+            chunk = fh.read(65536)
+            if not chunk:
+                break
+            for hash in six.viewvalues(toCompute):
+                hash.update(chunk)
+
+    fileModel.update({'_id': file['_id']}, update={
+        '$set': {alg: hash.hexdigest() for alg, hash in six.viewitems(toCompute)}
+    }, multi=False)
+
+
 def load(info):
     HashedFile(info['apiRoot'].file)
     ModelImporter.model('file').exposeFields(
-        level=AccessType.READ, fields=HashedFile.supportedAlgorithms)
+        level=AccessType.READ, fields=SUPPORTED_ALGORITHMS)
+
+    events.bind('data.process', info['name'], _computeHash)
```

Comment:
Moving forward, it would be nice to document what data are expected with which event ...